clients/TypeScript/packages/client/src/ChainSync/ChainSyncClient.ts

```diff
@@ -1,77 +1,88 @@
 import { Block, Ogmios, Point, Tip } from '@cardano-ogmios/schema'
 import {
   ConnectionConfig,
-  Mirror,
   InteractionContext,
   createClientContext
 } from '../Connection'
 import { UnknownResultError } from '../errors'
+import fastq from 'fastq'
 import { createPointFromCurrentTip, ensureSocketIsOpen } from '../util'
 import { findIntersect, Intersection } from './findIntersect'
 import { requestNext } from './requestNext'
 
 export interface ChainSyncClient {
   context: InteractionContext
-  requestNext: (options?: { mirror?: Mirror }) => void
   shutdown: () => Promise<void>
-  startSync: (points?: Point[], requestBuffer?: number) => Promise<Intersection>
+  startSync: (
+    points?: Point[],
+    inFlight?: number
+  ) => Promise<Intersection>
 }
 
 export interface ChainSyncMessageHandlers {
-  rollBackward: (response: {
-    point: Point,
-    tip: Tip,
-    reflection: Mirror
-  }) => void
-  rollForward: (response: {
-    block: Block,
-    tip: Tip,
-    reflection: Mirror
-  }) => void
+  rollBackward: (
+    response: {
+      point: Point,
+      tip: Tip
+    },
+    requestNext: () => void
+  ) => Promise<void>
+  rollForward: (
+    response: {
+      block: Block,
+      tip: Tip
+    },
+    requestNext: () => void
+  ) => Promise<void>
 }
 
 export const createChainSyncClient = async (
   messageHandlers: ChainSyncMessageHandlers,
   options?: {
-  connection?: ConnectionConfig
-}): Promise<ChainSyncClient> => {
+    connection?: ConnectionConfig,
+    sequential?: boolean
+  }): Promise<ChainSyncClient> => {
   return new Promise((resolve, reject) => {
     createClientContext(options).then(context => {
       const { socket } = context
+      const messageHandler = async (response: Ogmios['RequestNextResponse']) => {
+        if ('RollBackward' in response.result) {
+          await messageHandlers.rollBackward({
+            point: response.result.RollBackward.point,
+            tip: response.result.RollBackward.tip
+          }, () =>
+            requestNext(socket)
+          )
+        } else if ('RollForward' in response.result) {
+          await messageHandlers.rollForward({
+            block: response.result.RollForward.block,
+            tip: response.result.RollForward.tip
+          }, () => {
+            requestNext(socket)
+          })
+        } else {
+          throw new UnknownResultError(response.result)
+        }
+      }
+      const responseHandler = options?.sequential !== false
+        ? fastq.promise(messageHandler, 1).push
```

Comment:
That is one strange API, got me confused at first and had to look it up after looking for a corresponding `pop` in vain :sweat_smile: ...